Automatically deepen clones to find correct commits

RoystonS opened the pull request:
This fixes #494 - where we would get the wrong commit
that added a file if we were operating in a shallow
repository that didn't contain the actual commit -
by modifying `getCommitThatAddsFile` so that it
automatically deepens a shallow clone until it finds the
_real_ commit that added a file.

RoystonS commented:
(CLA requires some checking. Should be done soon.)

Note that I've hardwired the auto-deepening at 50 commits at a time.  It would be possible to make this configurable, or make it increase (say, double) the number of commits deepened each time around the loop.  But I'm not sure either are worthwhile. I'm imagining that one or two rounds of 50-deepening will cover the vast majority of scenarios.

Andarist commented on packages/git/src/index.test.ts:
```diff
@@ -189,16 +195,98 @@ describe("git", () => {
+        }
+      }
+
+      beforeEach(async () => {
```
I would prefer if those tests wouldn't use a shared `beforeEach` hook. It IMHO makes it harder to understand the exact flow of the test as it's not sufficient to read the test's body to understand all steps that are involved in it. With a help of helper functions, it should still be possible to write those 3 tests in a readable and somewhat DRY manner.

RoystonS replied:
Ok.  I'll see if I can pull out common helpers instead of a full common `beforeEach`.

Andarist commented on packages/git/src/index.ts:
```diff
@@ -43,13 +43,61 @@ export async function getDivergedCommit(cwd: string, ref: string) {
+    return isShallowResult.stdout.toString().trim() === "true";
+  }
+
+  async function deepenCloneBy(by: number) {
```
not yet sure how to best handle this but the code that consumes this starts parallel jobs:
https://github.com/atlassian/changesets/blob/c6d38b579bdfddc22714c352c3459d04d1745326/packages/apply-release-plan/src/index.ts#L203-L208
therefore this `getCommitThatAddsFile` can potentially be executed in parallel - while operating on a shared FS resource (git). It would be best to avoid this situation.

RoystonS replied:
Very good call - I hadn't appreciated the parallel usage.
Ok, so we need to avoid two things:
1. Multiple parallel calls unnecessarily each deepening a clone. (5 changesets not in the current clone shouldn't cause us to run 5x50-commit-deepen calls.)
2. Multiple parallel calls doing _anything_ to a repo (specifically the `git log --diff-filter`) whilst one call is deepening it. (Don't read a repo whilst writing to it.)

I'll add a (per-repo) mutex that allows multiple readers OR a single writer.

Testing the multiple parallel job behaviour will be an "interesting challenge".

Andarist replied:
> Testing the multiple parallel job behaviour will be an "interesting challenge".

I believe that sometimes there are just things not worth testing. Probably any test trying to test this would be rather strongly tied to the implementation and would require mocking or smth. Maybe its just such a case?

> I'll add a (per-repo) mutex that allows multiple readers OR a single writer. 

I would be interested in what do u mean by a single writer approach.

The alternative solution would be to provide a new function that would accept a list of files to check and return the list of commits. That way it could just implement a proper scheduling logic that would be local to it.

RoystonS replied:
Yes, testing such a beast fully would require some rather evil mocking and, as you say, be rather fragile and tied to implementation.

The single-writer approach would be some sort of mutex where as many things that want to read from the repo would be able to take 'read locks' as you like, but only one thing at a time can take a 'write lock' which blocks out readers until done.  Easy to make correct, but avoiding the multiple unnecessary deepenings would require some complex code.


A new function that accepts a simple list and can take control of its own scheduling would be _far_ simpler and more self-contained rather than having this specific git function having to know that it's being accessed in a parallel way and cope with that. i.e. the loop would be "Try to find commits for all files.  Any of them need a deepening? Ok, we'll do _one_, then repeat."

The only other use seems to be in the `apply-release-plan` test-utils: https://github.com/atlassian/changesets/blob/c6d38b579bdfddc22714c352c3459d04d1745326/packages/apply-release-plan/src/test-utils/get-changelog-entry-with-git-hash.ts#L8-L16

Ok, so I'll look into changing `getCommitThatAddsFile` so it's a bulk operation.  As we would want to encourage people to use the single bulk operation rather than looping, it probably makes sense to remove the existing one-at-a-time function, but that would be a breaking change in the `git` package. Let me know if that's not what you want.

Andarist replied:
If there is no conceptual problem with it we could make the implementation of the `getCommitThatAddsFile` as follows:
```js
const getCommitThatAddsFile = (file, cwd) => getCommitsThatAddsFiles([file], cwd)[0].commit
```

Making a breaking change to this package is not a big deal though (at least as far as I know) - it's just supposed to be used internally here and is not really up for public use.

Andarist reviewed (commented):
Overall it looks great and I'm confident that when we land this we'll improve the DX quite considerably. I've just left a comment about a single concern that I have - please take a look.

RoystonS commented:
Right. After a hectic release and a Christmas break I've gotten back to this and made the changes we discussed.

Specifically:

1. I've replaced `getCommitThatAddsFile` and its multiple parallel runs (which weren't safe when needing to deepen a shallow clone) with a single `getCommitsThatAddFiles` call that retrieve the commits in parallel and safely deepen a shallow clone.
2. I've removed the big `beforeEach` from the test to make the individual tests more standalone.

Note that I've left the existing `getCommitThatAddsFile` call in there (redirected through the new call) but have marked it as 'deprecated' using the NodeJS `deprecated` call.

RoystonS commented:
Ook, failing tests. Will investigate; they certainly pass locally.  In the meantime, code review very welcome!

Andarist commented on packages/git/src/index.ts:
```diff
@@ -1,8 +1,10 @@
 import { getPackages, Package } from "@manypkg/get-packages";
 import { GitError } from "@changesets/errors";
 import isSubdir from "is-subdir";
+import { deprecate } from "util";
```
TIL

RoystonS replied:
Nice little built-in API.  I haven't yet advertised the deprecation in the CHANGELOG entry.  I'll do that shortly.

RoystonS replied:
I've added a deprecation note to the changesets file

Andarist commented on packages/git/src/index.ts:
```diff
@@ -43,13 +45,126 @@ export async function getDivergedCommit(cwd: string, ref: string) {
+          commitsWithMissingParents.push(info);
+        }
+      } else {
+        // No commit for this file, which indicates it doesn't exist.
```
could you elaborate on what does it mean? the file? the commit? seems like a file - shouldn't we throw in such a case instead of silently ignoring this?

RoystonS replied:
This is the case where we've asked for the commit that introduced a file but the file doesn't exist.  It doesn't silently fail: it returns an `undefined` in that return slot.  So if you ask for the commits for 3 files, and only 2 exist, you'll still get the commits for the 2 that did exist, and `undefined` for the one that didn't exist. (There's a unit test for this.)

I could have made the bulk operation throw if any file didn't exist, but I think the previous implementation simply returned `undefined` for a non-existent file, so I kept that behaviour.  Happy to change it to a throw if any requested files don't exist; I slightly prefer that approach myself.

Andarist replied:
> I could have made the bulk operation throw if any file didn't exist, but I think the previous implementation simply returned undefined for a non-existent file, so I kept that behaviour. 

Right, good point about this being the current behavior. Let's keep it that way for now - but I think we might change this to throwing in the future.

Andarist commented on packages/git/src/index.ts:
```diff
@@ -43,13 +45,126 @@ export async function getDivergedCommit(cwd: string, ref: string) {
+      break;
+    }
+
+    remaining = commitsWithMissingParents.map(p => p.path);
```
this is only relevant for the `isRepoShallow()` case - so I would prefer keeping it in its block so the relation between them gets more obvious

RoystonS replied:
Yep, you're quite right. I'd obviously got a bit of tunnel vision in there.  That `remaining = ` line path only runs if we're deepening.  Incoming commit...

Andarist reviewed (commented):
I have left only some very minor commits - overall an excellent job 🚀

RoystonS commented:
Changes made. Back to you @Andarist when you get a moment.

Andarist reviewed (approved):
@RoystonS thank you for an excellent improvement and the hard work on this PR. I will be merging this 🔜

mitchellhamilton commented on packages/git/src/index.ts:
```diff
@@ -43,13 +45,125 @@ export async function getDivergedCommit(cwd: string, ref: string) {
+ * @param gitPaths - Paths to fetch
+ * @param cwd - Location of the repository
+ */
+async function getCommitsThatAddFiles(
```
Don't super mind if this happens in this PR but it might be nice if we batch/limit the concurrent deepens to 1 rather than having this because there could still be multiple calls to this in different places and etc.

RoystonS replied:
I'm a _little_ confused as to what you're asking for here when you say 'batch/limit rather than having this'.  This _is_ a batched API for deepening the repo (as discussed previously: https://github.com/atlassian/changesets/pull/495#discussion_r531805043).

I _could_ add a [mutex](https://www.npmjs.com/package/async-mutex) to `getCommitsThatAddFiles` if you're worried that there'd be multiple simultaneous calls to this batched API?

mitchellhamilton replied:
I mean keep the API getCommitThatAddsFile but batch it internally(probably just the deepen) and use a mutex or p-limit to limit simultaneous calls to deepen

RoystonS replied:
We discussed doing that previously (https://github.com/atlassian/changesets/pull/495#discussion_r531723754). Testing it would be a little tricky, hence @Andarist and I settling on the batch call.

I could rewrite and test it again, but could I ask you and @Andarist to jointly agree on an approach before I do so?

If you want to keep `getCommitThatAddsFile` and avoid introducing `getCommitsThatAddFiles` that should be done now before merging this PR otherwise doing it later will involve an undeprecation and another deprecation. :-(

mitchellhamilton replied:
Okay, happy to leave it how it is 👍
